File: algorithms/0046.permutations/permutations.rs
use std::mem::swap;

impl Solution {
    pub fn permute(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut ans = Vec::new();
        let mut nums = nums.clone();
        Self::backtracking(&mut nums, 0, &mut ans);
        ans
    }

    pub fn backtracking(nums: &mut Vec<i32>, level: usize, ans: &mut Vec<Vec<i32>>) {
        if level == nums.len() - 1 {
            ans.push(nums.to_vec());
            return;
        }

        for i in level..nums.len() {
            nums.swap(i, level);
            Self::backtracking(nums, level + 1, ans);
            nums.swap(i, level);
        }
    }

}

File: algorithms/0046.permutations/permutations.cpp
class Solution {
public:
    void backtracking(vector<int> &nums, int level, vector<vector<int>> &ans) {
        if (level == nums.size() -1) {
            ans.push_back(nums);
        }
        for (int i = level; i < nums.size(); ++i) {
            swap(nums[level], nums[i]);
            backtracking(nums, level + 1, ans);
            swap(nums[level], nums[i]);
        }
    }
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        backtracking(nums, 0, ans);
        return ans;
    }
};